### PR TITLE
SBRP Create Role Listing Form 

### PR DESCRIPTION
**Pull request by ZheTaoGuo**
### 📋 Overview (what is this PR about) 

Allows the SBRP admin to create role listing so that staff can apply for roles from other departments 

### 🚀 What are the features/changes included?
Role Listing form that has roles dropdown, description, start and end date fields and skills. 

### 🧪 What testing have I done to make sure it is working?
Upon form submission, backend receives post request with the role_name, start_date and end_date 

### 🧐 How should the reviewer go about testing it?
Start the backend and frontend, navigate to /create-role-listing and select the roles dropdown, description, start_date and end_date and submit the form, will redirect to the role-listings page

### ✅ Final check
- [ x] I have done the following:
  - `<git rebase main>` to replay your feature commits from the main branch
  - If there are any conflicts, resolve them then run:
    - `git add <resolved rebase conflict files>`
    - `<git rebase --continue>`
    - You may need to run `<git pull>` if your local feature branch is behind the remote feature branch
    - `<git push>`

### 🖼️ Make it visual ✨ (if possible, attach a screenshot of what the reviewer is supposed to see)
 
Screenshot 2023-09-29 at 5 20 23 PM

### ✍️ Any additional notes
https://spmg9t5.atlassian.net/jira/software/projects/T5/boards/1/backlog?selectedIssue=T5-4

### CODE REVIEW

File: backend/.idea/backend.iml
Comment (Jaylin0312) [inline backend/.idea/backend.iml:1]: `.idea` folder seems to be a folder generated by IntelliJ IDE, we can put this in `.gitignore`

File: frontend/src/routes.tsx [hunk 1]
```diff
@@ -11,6 +12,10 @@ const routes: RouteObject[] = [
     path: "all-role-listing",
     element: <AllRoleListing />,
   },
+  {
+    path: "create-role-listing",
+    element: <Signup />,
```
Comment (Jaylin0312) [inline frontend/src/routes.tsx]: Can change `<Signup />` to something more relevant eg. `<CreateRole />`

File: frontend/src/pages/hr/CreateRoleListing.tsx [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+import * as yup from "yup"
+
+import StaffNavbar from "../../components/Navbar"
+
+interface MyFormValues {
+  role_name: string
+  description: string
+  start_date: Dayjs | number | undefined
+  end_date: Dayjs | number | undefined
+}
```
Comment (Jaylin0312) [inline frontend/src/pages/hr/CreateRoleListing.tsx]: Consider moving this part to `types.d.ts`

File: frontend/src/pages/hr/CreateRoleListing.tsx [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  const [data, setData] = useState(null)
+  const [roles, setRoles] = useState<string[]>([])
+  const [retrievedSkills, setRetrievedSkills] = useState<string[]>([])
+  const [startDateValue, setstartDateValue] = useState<Dayjs | null>(null)
+  const [endDateValue, setendDateValue] = useState<Dayjs | null>(null)
+
+  useEffect(() => {
+    const rolesArray = new Array<string>()
+    const skillsSet = new Set<string>()
+    fetch(endpointUrl)
```
Comment (Jaylin0312) [inline frontend/src/pages/hr/CreateRoleListing.tsx:42]: Consider using `fetch("/api/roles")`. Might also be a good idea to move API logic to another file.

eg. Create an `api` folder & `roleAPI.py`, then put the logic inside

File: frontend/src/pages/hr/CreateRoleListing.tsx [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+      if (roleName === selectedRoleName) {
+        setRetrievedSkills(skills)
+      }
+    })
+  }
+  return (
+    <>
+      <StaffNavbar title={title} items={items} />
+      <Toaster />
+      <Grid container style={{ marginTop: "20px" }}>
```
Comment (Jaylin0312) [inline frontend/src/pages/hr/CreateRoleListing.tsx]: consider using tailwindcss styling here instead for consistency

**Review by Jaylin0312 — CHANGES_REQUESTED** [review]

**Comment by Jaylin0312** [comment]
CLGTM overall! Left some comments for the form component, can consider making the changes just to make the code cleaner + for future ease of maintenance if you have time.

Do also help to remove the `.idea` folder before we can merge this into main, thanks!

File: frontend/src/pages/hr/CreateRoleListing.tsx [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+import StaffNavbar from "../../components/Navbar"
+
+interface MyFormValues {
+  role_name: string
+  description: string
+  start_date: Dayjs | number | undefined
+  end_date: Dayjs | number | undefined
+}
+const signupSchema = yup.object().shape({
```
Comment (Jaylin0312) [inline frontend/src/pages/hr/CreateRoleListing.tsx]: change to `createRoleSchema` or something relevant

**Review by Jaylin0312 — CHANGES_REQUESTED** [review]

File: frontend/src/routes.tsx (hunk 1 above)
Comment (ZheTaoGuo) [reply frontend/src/routes.tsx]: Changed to CreateRoleListing

File: backend/.idea/backend.iml
Comment (ZheTaoGuo) [reply backend/.idea/backend.iml:1]: already in gitignore, removed the folder

File: frontend/src/pages/hr/CreateRoleListing.tsx (hunk 2 above)
Comment (ZheTaoGuo) [reply frontend/src/pages/hr/CreateRoleListing.tsx]: moved

File: frontend/src/pages/hr/CreateRoleListing.tsx (hunk 5 above)
Comment (ZheTaoGuo) [reply frontend/src/pages/hr/CreateRoleListing.tsx]: changed

**Comment by Jaylin0312** [comment]
Merging code into main

**Pull request merged**